### ISSUE TITLE
Segmentation Fault (SIGSEGV) in bool Unit::Killed()

### ISSUE THREAD

**Issue by stephengtuggy** [vegastrike/Vega-Strike-Engine-Source#11]
Just now, I got a segfault in the `inline bool` method `Unit::Killed()`. Judging by the comment just above, indicating that the `void Kill()` method deletes an object, it seems to me that trying to read the value of `Killed` is a bad idea in general. It would never return `true`, only `false` or... segfault. Why? Because in the latter case, the object has been deleted, and it no longer exists in memory!

We should probably do away with the whole concept of `Killed`, and use smart pointers instead, to avoid trying to access memory that has been reclaimed.

**Comment by stephengtuggy** [comment]
After further experimentation and examination of the code, it looks like we might need the `Killed` concept after all. However, to prevent other threads from trying to read that property on an object that has already been destroyed, I suggest using some form of thread-safe smart pointers. (See #8 ).

**Comment by stephengtuggy** [comment]
[std::atomic_flag](https://en.cppreference.com/w/cpp/atomic/atomic_flag) or similar might be worth using for this property, also.

**Comment by LifWirser** [comment]
While a "segfault" report is serious, this report is lacking enough information not to be

The minimum need for a proper BUG report is
- The type of os on which the error occured
- short description of machine  error occured on (64 bit ibm w nvidia-blah)
- error message displayed (segfault at blah.cpp ...)
- breif description of actions that caused problem
- is problem repeatable

Based on lack of needed informatiom I recommend this issue be closed

**Comment by stephengtuggy** [comment]
My apologies. I thought I had filled in the environment information for this bug months ago.

Okay, here goes;
The OS was Ubuntu 18.04 LTS, I believe.
64-bit
CPU: 8th-generation Intel Core i5-8500, 6-core, 3.00 GHz
Main system RAM: 16 GB of DDR4
Graphics: GeForce GTX 1050 Ti, with 4 GB of dedicated GDDR5
Motherboard: Gigabyte H370M-DS3H
Main system storage: 500GB M.2 NVMe SSD

I think there were two error messages displayed, one after the other. I no longer remember the first one, only the second one, which was the segfault message, in the terminal that I was running vegastrike from.

I do not have a specific series of actions to reproduce the problem, but I am guaranteed to see it on my hardware within about 20 minutes of gameplay or less, one way or another.

@LifWirser do you see anything else I'm missing?

**Comment by stephengtuggy** [comment]
This might still be an issue; I'm not sure. If it is, the problem may only show up when `DESTRUCTDEBUG` is defined.

**Comment by BenjamenMeyer** [comment]
Is this still an issue?

**Comment by stephengtuggy** [comment]
> Is this still an issue?

@BenjamenMeyer Yes, I believe so. In fact, there are multiple bugs in the unit-destruction code that can cause memory corruption and/or out-of-bounds access. I've done some work in this area, but haven't finished cleaning it all up yet.

To me, these bugs, and others of the same sort, are the highest priority to fix before we can ship the game. I suspect that there are literally hundreds of memory-management and multi-threading bugs in the code base currently. It won't be an easy task!
